File: backend/src/main/kotlin/com/mongle/service/PersonService.kt
package com.mongle.service

import com.mongle.common.Messages
import com.mongle.common.ValidationLimits
import com.mongle.common.Validators
import com.mongle.common.exception.BusinessException
import com.mongle.common.exception.ErrorCode
import com.mongle.controller.dto.PersonDetailResponse
import com.mongle.controller.dto.PersonRequest
import com.mongle.controller.dto.PersonResponse
import com.mongle.controller.dto.PersonSort
import com.mongle.domain.ChipType
import com.mongle.domain.Person
import com.mongle.domain.PersonRelationTag
import com.mongle.repository.ChipRepository
import com.mongle.repository.EventPersonRepository
import com.mongle.repository.EventRepository
import com.mongle.repository.PersonRelationTagRepository
import com.mongle.repository.PersonRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDate

@Service
@Transactional(readOnly = true)
class PersonService(
    private val personRepository: PersonRepository,
    private val chipService: ChipService,
    private val chipRepository: ChipRepository,
    private val personStatsService: PersonStatsService,
    private val eventRepository: EventRepository,
    private val personRelationTagRepository: PersonRelationTagRepository,
    private val eventPersonRepository: EventPersonRepository,
) {
    @Transactional
    fun register(userId: Long, request: PersonRequest): PersonResponse {
        val person = Person(ownerId = userId, name = request.name)
        val relationTagChipIds = applyRequest(userId, person, request)
        val saved = personRepository.save(person)
        syncRelationTags(requireNotNull(saved.id), relationTagChipIds)
        return toResponse(saved)
    }

    /** 전체 수정(PUT) — 등록과 같은 입력·검증을 재사용한다. 내 소유·active 인물만, 아니면 NOT_FOUND. */
    @Transactional
    fun update(userId: Long, personId: Long, request: PersonRequest): PersonResponse {
        val person = personRepository.findByIdAndOwnerIdAndDeletedAtIsNull(personId, userId)
            ?: throw BusinessException(ErrorCode.NOT_FOUND)
        val relationTagChipIds = applyRequest(userId, person, request)
        syncRelationTags(requireNotNull(person.id), relationTagChipIds)
        return toResponse(person)
    }

    /**
     * 디렉토리 목록(#29). 즐겨찾기를 항상 상단 그룹으로 두고 그 안에서 정렬한다:
     * 이름순(대소문자 무시) / 최근 만남순(마지막 만난 날 최신 먼저·없는 사람은 뒤).
     * 검색은 이름·관계 유형 부분 일치(대소문자 무시). 즐겨찾기 상단 조건이 있어 in-memory 로 조합한다.
     * (최근 만남순의 마지막 만난 날은 이후 파생 단계에서 event 반영으로 갱신된다.)
     */
    fun directory(userId: Long, sort: PersonSort, query: String?): List<PersonResponse> {
        val keyword = query?.trim()?.lowercase()?.ifBlank { null }
        val within = when (sort) {
            PersonSort.NAME -> compareBy(String.CASE_INSENSITIVE_ORDER) { p: Person -> p.name }
            PersonSort.RECENT -> Comparator.comparing(Person::lastMetDate, Comparator.nullsLast(Comparator.reverseOrder<LocalDate>()))
        }
        val order = compareByDescending<Person> { it.favorite }.then(within)
        val persons = personRepository.findByOwnerIdAndDeletedAtIsNull(userId)
            .filter { keyword == null || it.matches(keyword) }
            .sortedWith(order)
        // 관계태그·라벨을 한 번에 로드해 인물별 N+1 을 막는다.
        val tagChipIdsByPerson = relationTagChipIdsByPerson(persons)
        val tagLabels = resolveTagLabels(tagChipIdsByPerson.values.flatten())
        return persons.map { PersonResponse.from(it, tagChipIdsByPerson[it.id].orEmpty(), tagLabels) }
    }

    private fun Person.matches(keyword: String): Boolean = name.lowercase().contains(keyword) || relationType?.lowercase()?.contains(keyword) == true

    /** 상세 조회(#25). 기본 정보 + 파생 스탯(#30). 내 소유·active 인물만, 아니면 NOT_FOUND. */
    fun detail(userId: Long, personId: Long): PersonDetailResponse {
        val person = personRepository.findByIdAndOwnerIdAndDeletedAtIsNull(personId, userId)
            ?: throw BusinessException(ErrorCode.NOT_FOUND)
        val stats = personStatsService.statsOf(person)
        val tagChipIds = relationTagChipIdsOf(person)
        return PersonDetailResponse.from(person, stats, tagChipIds, resolveTagLabels(tagChipIds), LocalDate.now())
    }

    /** 즐겨찾기 토글(#28). 내 소유·active 인물만, 아니면 NOT_FOUND. */
    @Transactional
    fun toggleFavorite(userId: Long, personId: Long): PersonResponse {
        val person = personRepository.findByIdAndOwnerIdAndDeletedAtIsNull(personId, userId)
            ?: throw BusinessException(ErrorCode.NOT_FOUND)
        person.toggleFavorite()
        return toResponse(person)
    }

    /**
     * 삭제(#27). 인물을 소프트삭제하고, 연결된 active 기록마다 이 인물을 연결에서 제거한다 —
     * 그 결과 연결 인물이 0명이 되면(마지막 연결이었으면) 그 기록도 소프트삭제한다.
     * 다인 연결 기록은 통삭제하지 않는다(다른 사람 타임라인 보존, 판단 근거는 mustpass 02 §삭제).
     * 이미 소프트삭제된 기록은 findByPersonId 가 걸러 과거 참조를 보존한다.
     */
    @Transactional
    fun delete(userId: Long, personId: Long) {
        val person = personRepository.findByIdAndOwnerIdAndDeletedAtIsNull(personId, userId)
            ?: throw BusinessException(ErrorCode.NOT_FOUND)
        // active 기록에서만 이 인물 연결을 끊는다(소프트삭제된 기록은 findByPersonId 가 걸러 과거 참조 보존).
        // 연결이 0명이 된(마지막 연결이었던) 기록은 소프트삭제. 인물의 관계태그 행은 소프트삭제 시 남긴다(과거 참조 보존).
        eventRepository.findByPersonId(personId).forEach { event ->
            val eventId = requireNotNull(event.id)
            eventPersonRepository.deleteByEventIdAndPersonId(eventId, personId)
            if (eventPersonRepository.countByEventId(eventId) == 0L) event.softDelete()
        }
        person.softDelete()
    }

    /**
     * 등록·수정이 공유하는 입력 반영. 검증(글자수·날짜·태그·취향)을 모두 통과한 뒤에만 필드를 세팅한다.
     * 취향 목록은 엔티티에 바로 교체하고, 관계 태그는 조인 엔티티라 검증된 칩 id 목록만 돌려줘
     * 저장(id 확보) 후 호출자가 syncRelationTags 로 교체한다(PUT 시맨틱).
     */
    private fun applyRequest(userId: Long, person: Person, request: PersonRequest): List<Long> {
        val name = request.name.trim()
        Validators.requireNotBlank(name, Messages.REQUIRED_NAME)
        Validators.maxLength(name, ValidationLimits.NAME_MAX)

        val relationType = request.relationType?.trim()?.ifBlank { null }
        relationType?.let { Validators.maxLength(it, ValidationLimits.RELATION_TYPE_MAX) }

        val birthday = request.birthday
        PersonValidator.validateDates(
            birthYear = birthday?.year,
            birthMonth = birthday?.month,
            birthDay = birthday?.day,
            firstMetDate = request.firstMetDate,
            lastMetDate = request.lastMetDate,
        )

        // 관계태그 id 중복은 첫 등장 기준 1건으로 정규화한다 — 개수 상한도 정규화 이후 개수로 판단.
        val relationTagChipIds = request.relationTagChipIds.distinct()
        val allowedTagIds = chipService.visibleChips(userId, ChipType.RELATION_TAG).mapNotNull { it.id }.toSet()
        PersonValidator.validateRelationTags(relationTagChipIds, allowedTagIds)

        val likes = request.likes.map { it.trim() }.filter { it.isNotBlank() }
        val cautions = request.cautions.map { it.trim() }.filter { it.isNotBlank() }
        PersonValidator.validatePreferences(likes)
        PersonValidator.validatePreferences(cautions)

        // 월·일 없는 생일은 연도까지 무시한다(전부 null) — Birthday.from 이 월·일 기준으로 '생일 없음'을 판정하므로 연도만 남으면 응답에 안 보이는 유령 값이 된다.
        val effectiveBirthday = birthday?.takeIf { it.month != null && it.day != null }

        person.name = name
        person.birthYear = effectiveBirthday?.year
        person.birthMonth = effectiveBirthday?.month
        person.birthDay = effectiveBirthday?.day
        person.firstMetDate = request.firstMetDate
        person.lastMetDate = request.lastMetDate
        person.profileImageUrl = request.profileImageUrl?.trim()?.ifBlank { null }
        person.relationType = relationType
        person.favorite = request.favorite
        person.replaceLikes(likes)
        person.replaceCautions(cautions)
        return relationTagChipIds
    }

    /** 관계 태그 조인 행 전체 교체(수정 시 보낸 값으로 갈아끼움) — 하드삭제 후 순서대로 재삽입. */
    private fun syncRelationTags(personId: Long, chipIds: List<Long>) {
        personRelationTagRepository.deleteByPersonId(personId)
        chipIds.forEachIndexed { order, chipId ->
            personRelationTagRepository.save(PersonRelationTag(personId = personId, chipId = chipId, displayOrder = order))
        }
    }

    private fun relationTagChipIdsOf(person: Person): List<Long> = personRelationTagRepository
        .findByPersonIdOrderByDisplayOrderAsc(requireNotNull(person.id)).map { it.chipId }

    /** 여러 인물의 관계태그 칩 id 를 한 번에 로드(순서 보존) — 목록·홈이 재사용해 N+1 을 막는다. */
    fun relationTagChipIdsByPerson(persons: List<Person>): Map<Long, List<Long>> {
        val personIds = persons.mapNotNull { it.id }
        if (personIds.isEmpty()) return emptyMap()
        return personRelationTagRepository.findByPersonIdInOrderByPersonIdAscDisplayOrderAsc(personIds)
            .groupBy { it.personId }
            .mapValues { (_, rows) -> rows.map { it.chipId } }
    }

    /**
     * 관계 태그 라벨은 칩에서 해석한다 — id 참조라 이름이 바뀌면 자동 반영되고,
     * 소프트삭제된 칩도 findAllById 로 잡혀 라벨이 유지된다(과거 참조 보존, 00-infra).
     */
    private fun toResponse(person: Person): PersonResponse {
        val tagChipIds = relationTagChipIdsOf(person)
        return PersonResponse.from(person, tagChipIds, resolveTagLabels(tagChipIds))
    }

    private fun resolveTagLabels(chipIds: List<Long>): Map<Long, String> = chipRepository.findAllById(chipIds)
        .mapNotNull { chip -> chip.id?.let { it to chip.label } }
        .toMap()
}

File: backend/src/persons/person.service.ts
import { Prisma } from '@prisma/client';
import { Injectable } from '@nestjs/common';
import { BusinessException } from '../common/exception/business-exception';
import { ErrorCode } from '../common/exception/error-code';
import { toNumberId } from '../common/prisma';
import { ChipService } from '../chips/chip.service';
import { ChipType } from '../chips/chip.dto';
import { PrismaService } from '../prisma/prisma.service';
import { ChipRef } from '../shared/api.dto';
import {
  acquaintancePeriod,
  compareDates,
  daysSinceFirstMet,
  formatDate,
  formatLocalDateTime,
  nowAsKstLocalDateTime,
  parseDate,
  relativeDate,
  todayInKst,
} from '../shared/date';
import {
  LIMITS,
  MESSAGES,
  booleanOrDefault,
  integerIds,
  invalidInput,
  maxLength,
  notFuture,
  optionalDate,
  optionalText,
  requiredText,
  stringList,
  validDateOrder,
} from '../shared/validation';
import {
  Birthday,
  PersonDetailResponse,
  PersonGender,
  PersonRequest,
  PersonResponse,
  PersonSort,
  PersonStats,
} from './person.dto';

type PersonRow = {
  id: bigint;
  ownerId: Uint8Array<ArrayBuffer>;
  name: string;
  birthYear: number | null;
  birthMonth: number | null;
  birthDay: number | null;
  firstMetDate: Date | null;
  lastMetDate: Date | null;
  profileImageUrl: string | null;
  gender: string | null;
  relationType: string | null;
  favorite: boolean;
  createdAt: Date | null;
  updatedAt: Date | null;
  deletedAt: Date | null;
};

type NormalizedPerson = {
  name: string;
  birthday: { year: number | null; month: number; day: number } | null;
  firstMetDate: string | null;
  lastMetDate: string | null;
  profileImageUrl: string | null;
  gender: PersonGender | null;
  relationType: string | null;
  relationTagChipIds: bigint[];
  likes: string[];
  cautions: string[];
  favorite: boolean;
};

export type PersonStatsData = {
  meetingDatesDesc: string[];
  recordCount: number;
  lastMetDate: string | null;
};

@Injectable()
export class PersonService {
  constructor(
    private readonly prisma: PrismaService,
    private readonly chips: ChipService,
  ) {}

  async register(ownerId: Buffer<ArrayBuffer>, request: PersonRequest): Promise<PersonResponse> {
    const normalized = await this.normalize(ownerId, request);
    const now = nowAsKstLocalDateTime();
    const person = await this.prisma.$transaction(async (tx) => {
      const created = await tx.person.create({
        data: {
          ownerId,
          name: normalized.name,
          birthYear: normalized.birthday?.year ?? null,
          birthMonth: normalized.birthday?.month ?? null,
          birthDay: normalized.birthday?.day ?? null,
          firstMetDate: normalized.firstMetDate ? parseDate(normalized.firstMetDate) : null,
          lastMetDate: normalized.lastMetDate ? parseDate(normalized.lastMetDate) : null,
          profileImageUrl: normalized.profileImageUrl,
          gender: normalized.gender,
          relationType: normalized.relationType,
          favorite: normalized.favorite,
          createdAt: now,
          updatedAt: now,
        },
      });
      await this.replacePersonCollections(tx, created.id, normalized, now);
      return created;
    });
    return this.toResponse(person);
  }

  async update(ownerId: Buffer<ArrayBuffer>, personId: bigint, request: PersonRequest): Promise<PersonResponse> {
    await this.loadOwned(ownerId, personId);
    const normalized = await this.normalize(ownerId, request);
    const now = nowAsKstLocalDateTime();
    const person = await this.prisma.$transaction(async (tx) => {
      const updated = await tx.person.update({
        where: { id: personId },
        data: {
          name: normalized.name,
          birthYear: normalized.birthday?.year ?? null,
          birthMonth: normalized.birthday?.month ?? null,
          birthDay: normalized.birthday?.day ?? null,
          firstMetDate: normalized.firstMetDate ? parseDate(normalized.firstMetDate) : null,
          lastMetDate: normalized.lastMetDate ? parseDate(normalized.lastMetDate) : null,
          profileImageUrl: normalized.profileImageUrl,
          gender: normalized.gender,
          relationType: normalized.relationType,
          favorite: normalized.favorite,
          updatedAt: now,
        },
      });
      await this.replacePersonCollections(tx, personId, normalized, now);
      return updated;
    });
    return this.toResponse(person);
  }

  async directory(ownerId: Buffer<ArrayBuffer>, sort: PersonSort, query?: string): Promise<PersonResponse[]> {
    const keyword = query?.trim().toLowerCase() || null;
    const persons = (await this.prisma.person.findMany({ where: { ownerId, deletedAt: null } })) as PersonRow[];
    const filtered = persons.filter(
      (person) =>
        keyword === null ||
        person.name.toLowerCase().includes(keyword) ||
        person.relationType?.toLowerCase().includes(keyword) === true,
    );
    filtered.sort((left, right) => {
      if (left.favorite !== right.favorite) return left.favorite ? -1 : 1;
      if (sort === PersonSort.RECENT) {
        if (left.lastMetDate === null && right.lastMetDate !== null) return 1;
        if (left.lastMetDate !== null && right.lastMetDate === null) return -1;
        if (left.lastMetDate !== null && right.lastMetDate !== null) {
          const recent = compareDates(formatDate(right.lastMetDate), formatDate(left.lastMetDate));
          if (recent !== 0) return recent;
        }
        return 0;
      }
      return compareText(left.name.toLowerCase(), right.name.toLowerCase());
    });
    // 목록 크기와 무관하게 태그·취향 컬렉션을 한 번씩만 읽는다.
    return this.toResponses(filtered);
  }

  async detail(ownerId: Buffer<ArrayBuffer>, personId: bigint): Promise<PersonDetailResponse> {
    const person = await this.loadOwned(ownerId, personId);
    const [base, stats] = await Promise.all([this.toResponse(person), this.statsOf(person)]);
    const today = todayInKst();
    const responseStats: PersonStats = {
      meetCount: stats.meetingDatesDesc.length,
      recordCount: stats.recordCount,
      daysSinceFirstMet: base.firstMetDate ? daysSinceFirstMet(base.firstMetDate, today) : null,
      acquaintancePeriod: base.firstMetDate ? acquaintancePeriod(base.firstMetDate, today) : null,
      lastMetRelative: stats.lastMetDate ? relativeDate(stats.lastMetDate, today) : null,
    };
    return { ...base, lastMetDate: stats.lastMetDate, stats: responseStats };
  }

  async toggleFavorite(ownerId: Buffer<ArrayBuffer>, personId: bigint): Promise<PersonResponse> {
    const person = await this.loadOwned(ownerId, personId);
    const updated = await this.prisma.person.update({
      where: { id: personId },
      data: { favorite: !person.favorite, updatedAt: nowAsKstLocalDateTime() },
    });
    return this.toResponse(updated as PersonRow);
  }

  async delete(ownerId: Buffer<ArrayBuffer>, personId: bigint): Promise<void> {
    await this.loadOwned(ownerId, personId);
    const now = nowAsKstLocalDateTime();
    await this.prisma.$transaction(async (tx) => {
      const links = await tx.eventPerson.findMany({ where: { personId } });
      const eventIds = links.map((link) => link.eventId);
      const activeEvents = eventIds.length
        ? await tx.event.findMany({ where: { id: { in: eventIds }, deletedAt: null } })
        : [];
      for (const event of activeEvents) {
        await tx.eventPerson.deleteMany({ where: { eventId: event.id, personId } });
        const remaining = await tx.eventPerson.count({ where: { eventId: event.id } });
        if (remaining === 0) {
          await tx.event.update({ where: { id: event.id }, data: { deletedAt: now, updatedAt: now } });
        }
      }
      await tx.person.update({ where: { id: personId }, data: { deletedAt: now, updatedAt: now } });
    });
  }

  async loadOwned(ownerId: Buffer<ArrayBuffer>, personId: bigint): Promise<PersonRow> {
    const person = await this.prisma.person.findFirst({ where: { id: personId, ownerId, deletedAt: null } });
    if (!person) throw new BusinessException(ErrorCode.NOT_FOUND);
    return person as PersonRow;
  }

  async relationTagChipIdsByPerson(persons: PersonRow[]): Promise<Map<bigint, bigint[]>> {
    const ids = persons.map((person) => person.id);
    if (ids.length === 0) return new Map();
    const rows = await this.prisma.personRelationTag.findMany({
      where: { personId: { in: ids } },
      orderBy: [{ personId: 'asc' }, { displayOrder: 'asc' }],
    });
    const result = new Map<bigint, bigint[]>();
    for (const row of rows) {
      const values = result.get(row.personId) ?? [];
      values.push(row.chipId);
      result.set(row.personId, values);
    }
    return result;
  }

  async statsOf(person: PersonRow): Promise<PersonStatsData> {
    const meetingCategoryId = await this.chips.meetingCategoryId();
    const links = await this.prisma.eventPerson.findMany({ where: { personId: person.id } });
    const eventIds = links.map((link) => link.eventId);
    const events = eventIds.length
      ? await this.prisma.event.findMany({ where: { id: { in: eventIds }, deletedAt: null } })
      : [];
    const meetingDatesDesc = meetingCategoryId
      ? [
          ...new Set(
            events
              .filter((event) => event.categoryChipId === meetingCategoryId)
              .map((event) => formatDate(event.occurredDate)),
          ),
        ].sort((left, right) => compareDates(right, left))
      : [];
    const candidates = [person.lastMetDate ? formatDate(person.lastMetDate) : null, meetingDatesDesc[0] ?? null]
      .filter((value): value is string => value !== null)
      .filter((value) => person.firstMetDate === null || compareDates(value, formatDate(person.firstMetDate)) >= 0)
      .sort((left, right) => compareDates(right, left));
    return { meetingDatesDesc, recordCount: events.length, lastMetDate: candidates[0] ?? null };
  }

  private async normalize(ownerId: Buffer<ArrayBuffer>, request: PersonRequest): Promise<NormalizedPerson> {
    if (!request || typeof request !== 'object' || Array.isArray(request)) return invalidInput();
    const name = requiredText(request.name, MESSAGES.requiredName);
    maxLength(name, LIMITS.name);
    const relationType = optionalText(request.relationType);
    if (relationType !== null) maxLength(relationType, LIMITS.relationType);

    const birthday = normalizeBirthday(request.birthday);
    const firstMetDate = optionalDate(request.firstMetDate);
    const lastMetDate = optionalDate(request.lastMetDate);
    if (birthday?.year !== null && birthday !== null) notFuture(dateFromBirthday(birthday));
    if (firstMetDate !== null) notFuture(firstMetDate);
    if (lastMetDate !== null) notFuture(lastMetDate);
    validDateOrder(firstMetDate, lastMetDate);

    const relationTagChipIds = integerIds(request.relationTagChipIds);
    if (relationTagChipIds.length > LIMITS.relationTags) {
      throw new BusinessException(ErrorCode.SELECTION_LIMIT, MESSAGES.relationTagLimit);
    }
    const visibleTagIds = new Set(
      (await this.chips.visibleChips(ownerId, ChipType.RELATION_TAG)).map((chip) => chip.id.toString()),
    );
    if (relationTagChipIds.some((id) => !visibleTagIds.has(id.toString())))
      throw new BusinessException(ErrorCode.NOT_FOUND);

    const likes = normalizePreferences(request.likes);
    const cautions = normalizePreferences(request.cautions);
    const profileImageUrl = optionalText(request.profileImageUrl);
    const gender = normalizeGender(request.gender);
    const favorite = booleanOrDefault(request.favorite, false);
    return {
      name,
      birthday,
      firstMetDate,
      lastMetDate,
      profileImageUrl,
      gender,
      relationType,
      relationTagChipIds,
      likes,
      cautions,
      favorite,
    };
  }

  private async replacePersonCollections(
    tx: Prisma.TransactionClient,
    personId: bigint,
    normalized: NormalizedPerson,
    now: Date,
  ): Promise<void> {
    await Promise.all([
      tx.personRelationTag.deleteMany({ where: { personId } }),
      tx.personLike.deleteMany({ where: { personId } }),
      tx.personCaution.deleteMany({ where: { personId } }),
    ]);
    if (normalized.relationTagChipIds.length) {
      await tx.personRelationTag.createMany({
        data: normalized.relationTagChipIds.map((chipId, displayOrder) => ({
          personId,
          chipId,
          displayOrder,
          createdAt: now,
          updatedAt: now,
        })),
      });
    }
    if (normalized.likes.length) {
      await tx.personLike.createMany({
        data: normalized.likes.map((item, itemOrder) => ({ personId, itemOrder, item })),
      });
    }
    if (normalized.cautions.length) {
      await tx.personCaution.createMany({
        data: normalized.cautions.map((item, itemOrder) => ({ personId, itemOrder, item })),
      });
    }
  }

  private async toResponse(person: PersonRow): Promise<PersonResponse> {
    return (await this.toResponses([person]))[0];
  }

  private async toResponses(persons: PersonRow[]): Promise<PersonResponse[]> {
    if (persons.length === 0) return [];
    const personIds = persons.map((person) => person.id);
    const [tagRows, likeRows, cautionRows] = await Promise.all([
      this.prisma.personRelationTag.findMany({
        where: { personId: { in: personIds } },
        orderBy: [{ personId: 'asc' }, { displayOrder: 'asc' }],
      }),
      this.prisma.personLike.findMany({
        where: { personId: { in: personIds } },
        orderBy: [{ personId: 'asc' }, { itemOrder: 'asc' }],
      }),
      this.prisma.personCaution.findMany({
        where: { personId: { in: personIds } },
        orderBy: [{ personId: 'asc' }, { itemOrder: 'asc' }],
      }),
    ]);
    const tagIdsByPerson = groupByPerson(tagRows, (row) => row.chipId);
    const likesByPerson = groupByPerson(likeRows, (row) => row.item ?? '');
    const cautionsByPerson = groupByPerson(cautionRows, (row) => row.item ?? '');
    const tagIds = [...new Set(tagRows.map((row) => row.chipId))];
    const chips = tagIds.length ? await this.prisma.chip.findMany({ where: { id: { in: tagIds } } }) : [];
    const chipsById = new Map(chips.map((chip) => [chip.id.toString(), chip]));
    return persons.map((person) => {
      const relationTags: ChipRef[] = (tagIdsByPerson.get(person.id) ?? []).flatMap((id) => {
        const chip = chipsById.get(id.toString());
        return chip ? [{ id: toNumberId(id), label: chip.label, color: chip.color }] : [];
      });
      return {
        id: toNumberId(person.id),
        name: person.name,
        birthday: birthdayFrom(person),
        firstMetDate: person.firstMetDate ? formatDate(person.firstMetDate) : null,
        lastMetDate: person.lastMetDate ? formatDate(person.lastMetDate) : null,
        profileImageUrl: person.profileImageUrl,
        gender: person.gender as PersonGender | null,
        relationType: person.relationType,
        relationTags,
        likes: likesByPerson.get(person.id) ?? [],
        cautions: cautionsByPerson.get(person.id) ?? [],
        favorite: person.favorite,
        createdAt: formatLocalDateTime(person.createdAt),
      };
    });
  }
}

function compareText(left: string, right: string): number {
  return left < right ? -1 : left > right ? 1 : 0;
}

function normalizeGender(value: unknown): PersonGender | null {
  if (value === undefined || value === null) return null;
  if (value !== PersonGender.FEMALE && value !== PersonGender.MALE) return invalidInput();
  return value;
}

function normalizeBirthday(value: unknown): NormalizedPerson['birthday'] {
  if (value === undefined || value === null) return null;
  if (typeof value !== 'object' || Array.isArray(value)) return invalidInput();
  const input = value as Record<string, unknown>;
  const year = nullableInteger(input.year);
  const month = nullableInteger(input.month);
  const day = nullableInteger(input.day);
  if (month === null && day === null) return null;
  if (month === null || day === null) return invalidInput();
  const probe = `${String(year ?? 2000).padStart(4, '0')}-${String(month).padStart(2, '0')}-${String(day).padStart(2, '0')}`;
  try {
    parseDate(probe);
  } catch {
    return invalidInput();
  }
  return { year, month, day };
}

function nullableInteger(value: unknown): number | null {
  if (value === undefined || value === null) return null;
  if (typeof value !== 'number' || !Number.isInteger(value)) return invalidInput();
  return value;
}

function dateFromBirthday(birthday: NonNullable<NormalizedPerson['birthday']>): string {
  return `${String(birthday.year).padStart(4, '0')}-${String(birthday.month).padStart(2, '0')}-${String(birthday.day).padStart(2, '0')}`;
}

function normalizePreferences(value: unknown): string[] {
  const items = stringList(value)
    .map((item) => item.trim())
    .filter(Boolean);
  if (items.length > LIMITS.preferences) {
    throw new BusinessException(ErrorCode.SELECTION_LIMIT, MESSAGES.preferenceLimit);
  }
  for (const item of items) maxLength(item, LIMITS.preferenceItem);
  if (new Set(items).size !== items.length) throw new BusinessException(ErrorCode.DUPLICATE);
  return items;
}

function birthdayFrom(person: PersonRow): Birthday | null {
  if (person.birthMonth === null || person.birthDay === null) return null;
  return { year: person.birthYear, month: person.birthMonth, day: person.birthDay };
}

function groupByPerson<Row extends { personId: bigint }, Value>(
  rows: Row[],
  valueOf: (row: Row) => Value,
): Map<bigint, Value[]> {
  const result = new Map<bigint, Value[]>();
  for (const row of rows) {
    const values = result.get(row.personId) ?? [];
    values.push(valueOf(row));
    result.set(row.personId, values);
  }
  return result;
}
